binary_search_tree/tree.py

```diff
@@ -7,52 +7,129 @@ def __init__(self, key, val = None):
+
+        return current_node
+
+    # # Time Complexity: O(log n)
+    # # Space Complexity: O(1)
+    def add(self, key, value=None):
+        node = TreeNode(key, value)
+        if self.root == None:
+            self.root = node
+            return
+
+        self.add_helper(self.root, node)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def find_helper(self, current, key):
+        if current == None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif key < current.key:
+            return self.find_helper(current.left, key)
+
+        return self.find_helper(current.right, key)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, current, traversal_list):
+        if current != None:
+            self.inorder_helper(current.left, traversal_list)
+            traversal_list.append({'key': current.key, 'value': current.value})
+            self.inorder_helper(current.right, traversal_list)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def inorder(self):
-        pass
+        traversal_list = []
+        self.inorder_helper(self.root, traversal_list)
+        return traversal_list
+
+    def preorder_helper(self, current, traversal_list):
+        if current != None:
+            traversal_list.append({'key': current.key, 'value': current.value})
+            self.preorder_helper(current.left, traversal_list)
+            self.preorder_helper(current.right, traversal_list)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def preorder(self):
-        pass
+        traversal_list = []
+        self.preorder_helper(self.root, traversal_list)
+        return traversal_list
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current, traversal_list):
+        if current != None:
+            self.postorder_helper(current.left, traversal_list)
+            self.postorder_helper(current.right, traversal_list)
+            traversal_list.append({'key': current.key, 'value': current.value})
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def postorder(self):
-        pass
+        traversal_list = []
+        self.postorder_helper(self.root, traversal_list)
+        return traversal_list
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+    def height_helper(self, current):
+        if current == None:
+            return 0
+        left = self.height_helper(current.left)
+        right = self.height_helper(current.right)
 
+        return max(left, right) + 1
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
+    def height(self):
```

Comment:
👍 But the space complexity is O(log n) for a balanced tree due to the call stack (recursion)